### ISSUE TITLE
Extremely high memory usage for continuous models

### ISSUE THREAD

**Issue by Tortar** [JuliaDynamics/Agents.jl#739]
The memory usage for the `flockers` models is way too high, I didn't look carefully to see where this goes wrong (probably some unbounded cache) but it is something which surely requires work. For the code below it has a pick of more than 8 gb of memory. 

<details> <summary> Flocker model benchmark </summary>

```julia
using Agents, LinearAlgebra

@agent Bird ContinuousAgent{2} begin
    speed::Float64
    cohere_factor::Float64
    separation::Float64
    separate_factor::Float64
    match_factor::Float64
    visual_distance::Float64
end

function initialize_model(;
    n_birds = 3200,
    speed = 1.0,
    cohere_factor = 0.25,
    separation = 4.0,
    separate_factor = 0.25,
    match_factor = 0.01,
    visual_distance = 10.0,
    extent = (200, 200),
)
    space2d = ContinuousSpace(extent; spacing = 1)
    model = ABM(Bird, space2d, scheduler = Schedulers.Randomly())
    for _ in 1:n_birds
        vel = Tuple(rand(model.rng, 2) * 2 .- 1)
        add_agent!(
            model,
            vel,
            speed,
            cohere_factor,
            separation,
            separate_factor,
            match_factor,
            visual_distance,
        )
    end
    return model
end
nothing # hide

# ## Defining the agent_step!
# `agent_step!` is the primary function called for each step and computes velocity
# according to the three rules defined above.
function agent_step!(bird, model)
    ## Obtain the ids of neighbors within the bird's visual distance
    neighbor_ids = nearby_ids(bird, model, bird.visual_distance)
    N = 0
    match = separate = cohere = (0.0, 0.0)
    ## Calculate behaviour properties based on neighbors
    for id in neighbor_ids
        N += 1
        neighbor = model[id].pos
        heading = neighbor .- bird.pos

        ## `cohere` computes the average position of neighboring birds
        cohere = cohere .+ heading
        if euclidean_distance(bird.pos, neighbor, model) < bird.separation
            ## `separate` repels the bird away from neighboring birds
            separate = separate .- heading
        end
        ## `match` computes the average trajectory of neighboring birds
        match = match .+ model[id].vel
    end
    N = max(N, 1)
    ## Normalise results based on model input and neighbor count
    cohere = cohere ./ N .* bird.cohere_factor
    separate = separate ./ N .* bird.separate_factor
    match = match ./ N .* bird.match_factor
    ## Compute velocity based on rules defined above
    bird.vel = (bird.vel .+ cohere .+ separate .+ match) ./ 2
    bird.vel = bird.vel ./ norm(bird.vel)
    ## Move bird according to new velocity and speed
    move_agent!(bird, model, bird.speed)

end


model = initialize_model()
step!(model, agent_step!, 200)
```
</details>

It even invalidates some frameworks comparisons [here](https://github.com/isislab-unisa/ABM_Comparison)

**Comment by Tortar** [comment]
yes, the problem seems to reside in the fact that `offsets_within_radius` is a dictionary with arbitrary float values as keys, the problem with that is that it continues to grow over time in continuous spaces. Just took a quick look to internals, so I could be wrong, hope it helps

**Comment by Datseris** [comment]
Yes that is correct, the caching happens because (typically) the radius used in the search would be the same, which saves the time of re-creating the offset vectors for each search radius. It is also true that you would save memory if you did not cache the offset vectors for the radius. 

What is not clear to me is whether this would save time. Would the non-saving-cache version faster in the end...? Needs to be benchmarked.

Thankfully it is easy to add a simple option `cache_neighbor_searches = true` that if false there is no cache saving and the offset vectors are created on the spot each time the function is called.

**Comment by Datseris** [comment]
Wait a moment, in your example the birds all have the same vision horizon...? So then the dictionary only allocates once and this is not the problem. Still, probably a good idea to make a PR with the option to not save cache.

**Comment by Tortar** [comment]
yes, that is correct, all have the same vision

**Comment by Datseris** [comment]
Okay, so can you provide the numbers here? What do you mean by "there is 8 gb memory"?

**Comment by Datseris** [comment]
Where is this memory, what is the peak. Is it the CPU memory you measure externally, is it the memory reported by `@time`?

**Comment by Datseris** [comment]
This makes a huge difference: the _total_ allocated memory can be as high as it wants, the question is what is the maximum allocated memory at a given point in time. This surely should not be 8 gb, if it is then there is a bug.

**Comment by Tortar** [comment]
maximum alllocated memory, the peak, calculated looking at thememory monitor

**Comment by Datseris** [comment]
There is a tool that is in VSCode, and is called `@profview_allocs`. Here is how to use it (replace `@profview` with `@profview_allocs`) https://www.julia-vscode.org/docs/stable/userguide/profiler/  

It provides profiling information on the allocations. You could use this tool on a loop over e.g., 1000 calls to `step!`.

**Comment by Tortar** [comment]
Will not be able to create a memgraph on VSCode unfortunately (it doesn't show up), but I will add to the PR the screens of my system monitor at the end of the simulation to show what happens

**Comment by Tortar** [comment]
Solved in #740
